projects/plutus/src/ltc_afe.c

```diff
@@ -0,0 +1,286 @@
+  LTC6804_RDCVB_RESERVED,
+  LTC6804_RDCVC_RESERVED,
+  LTC6804_RDCVD_RESERVED,
+  LTC6804_RDAUXA_RESERVED,
+  LTC6804_RDAUXA_RESERVED,
+  LTC6804_RDSTATA_RESERVED,
+  LTC6804_RDSTATB_RESERVED,
+  LTC6804_RDCOMM_RESERVED
+};
+
+static uint8_t s_voltage_reg[NUM_LTC_AFE_VOLTAGE_REGISTER] = {
+  LTC_AFE_REGISTER_CELL_VOLTAGE_A,
+  LTC_AFE_REGISTER_CELL_VOLTAGE_B,
+  LTC_AFE_REGISTER_CELL_VOLTAGE_C,
+  LTC_AFE_REGISTER_CELL_VOLTAGE_D
+};
+
+static void prv_wakeup_idle(const LTCAFESettings *afe) {
+  gpio_set_pin_state(&afe->cs, GPIO_STATE_LOW);
+  delay_us(2);
+  gpio_set_pin_state(&afe->cs, GPIO_STATE_HIGH);
+}
+
+static void prv_build_cmd(uint16_t command, uint8_t *cmd) {
+  cmd[0] = (uint8_t)(command >> 8);
+  cmd[1] = (uint8_t)(command & 0xFF);
+
+  uint16_t cmd_pec = crc15_calculate(cmd, 2);
+  cmd[2] = (uint8_t)(cmd_pec >> 8);
+  cmd[3] = (uint8_t)(cmd_pec);
+}
+
+static StatusCode prv_read_register(const LTCAFESettings *afe,
+                                    LTCAFERegister reg, uint8_t *data, size_t len) {
+  if (reg > NUM_LTC_AFE_REGISTER) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
+  uint16_t reg_cmd = s_read_reg_cmd[reg];
+
+  uint8_t cmd[4] = { 0 };
+  prv_build_cmd(reg_cmd, cmd);
+
+  prv_wakeup_idle(afe);
+  return spi_exchange(afe->spi_port, cmd, 4, data, len);
+}
+
+// read from a voltage register
+static StatusCode prv_read_voltage(LTCAFESettings *afe,
+                                    LTCAFEVoltageRegister reg, LTCAFEVoltageRegisterGroup *data) {
+  if (reg > NUM_LTC_AFE_VOLTAGE_REGISTER) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
+  size_t len = sizeof(LTCAFEVoltageRegisterGroup) * PLUTUS_AFE_DEVICES_IN_CHAIN;
+  return prv_read_register(afe, s_voltage_reg[reg], (uint8_t *)data, len);
+}
+
+// start cell voltage conversion
+static void prv_trigger_adc_conversion(const LTCAFESettings *afe) {
+  uint8_t mode = (uint8_t)((afe->adc_mode + 1) % 3);
+  // ADCV command
+  uint16_t adcv = LTC6804_ADCV_RESERVED | LTC6804_ADCV_DISCHARGE_NOT_PERMITTED
+                  | LTC6804_CNVT_CELL_ALL | (mode << 7);
+
+  uint8_t cmd[4] = { 0 };
+  prv_build_cmd(adcv, cmd);
+
+  prv_wakeup_idle(afe);
+  spi_exchange(afe->spi_port, cmd, 4, NULL, 0);
+
+  // wait for conversions to finish
+  delay_ms(100);
+}
+
+static void prv_trigger_aux_adc_conversion(const LTCAFESettings *afe) {
+  uint8_t mode = (uint8_t)((afe->adc_mode + 1) % 3);
+  // ADAX
+  uint16_t adax = LTC6804_ADAX_RESERVED | LTC6804_ADAX_GPIO1 | (mode << 7);
+
+  uint8_t cmd[4] = { 0 };
+  prv_build_cmd(adax, cmd);
+
+  prv_wakeup_idle(afe);
+  spi_exchange(afe->spi_port, cmd, 4, NULL, 0);
+
+  // wait for conversions to finish
+  delay_ms(10);
+}
+
+// write config to all devices
+static StatusCode prv_write_config(const LTCAFESettings *afe, uint8_t gpio_enable_pins) {
+  // see p.54 in datasheet
+  LTCAFEWriteConfigPacket config_packet = { 0 };
+
+  // WRCFG
+  config_packet.wrcfg.data_hi = (uint8_t)(LTC6804_WRCFG_RESERVED >> 8);
```

Comment:
Guess not. I'll change it to use ``prv_build_cmd`` and allocate 4 bytes like the other ones